### PR TITLE
[BEAM-23836] Updating documentation for cross-language Java pipelines…

### PR DESCRIPTION
**Pull request by ihji**
… using Python external transforms

#23836

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by ihji** [comment]
R: @pcoet @chamikaramj

File: website/www/site/content/en/documentation/programming-guide.md [hunk 1]
```diff
@@ -7393,7 +7402,29 @@ Depending on the SDK language of the pipeline, you can use a high-level SDK-wrap
...
+Users have three options to use cross-language transforms in a Java pipeline. At the highest level of abstraction, some popular Python transforms are accessible through dedicated Java wrapper transforms. For example, Java SDK has `DataframeTransform` class which uses Python SDK's `DataframeTransform` and `RunInference` class in Java SDK for `RunInference` in Python SDK and so on. When a SDK-specific wrapper transform is not available for a target Python transform, you could use the lower-level [PythonExternalTransform](https://github.com/apache/beam/blob/master/sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/PythonExternalTransform.java) class instead by specifying the fully qualified name of the Python transform. If you want to try external transforms from the SDKS other than Python, you may also use the lowest-level [External](https://github.com/apache/beam/blob/master/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java) class.
```
Comment (pcoet) [inline website/www/site/content/en/documentation/programming-guide.md:7405]: Lightly edited. Please make sure this still says what you mean:
```
Users have three options to use cross-language transforms in a Java pipeline. At the highest level of abstraction, some popular Python transforms are accessible through dedicated Java wrapper transforms. For example, the Java SDK has the `DataframeTransform` class, which uses the Python SDK's `DataframeTransform`, and it has the `RunInference` class, which uses the Python SDK's `RunInference`, and so on. When an SDK-specific wrapper transform is not available for a target Python transform, you can use the lower-level [PythonExternalTransform](https://github.com/apache/beam/blob/master/sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/PythonExternalTransform.java) class instead by specifying the fully qualified name of the Python transform. If you want to try external transforms from SDKs other than Python, you can also use the lowest-level [External](https://github.com/apache/beam/blob/master/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java) class.
```

File: website/www/site/content/en/documentation/programming-guide.md [hunk 2]
```diff
@@ -7393,7 +7402,29 @@ Depending on the SDK language of the pipeline, you can use a high-level SDK-wrap
...
+
+```java
+import org.apache.beam.sdk.extensions.python.transforms.DataframeTransform;
+
+input.apply(DataframeTransform.of("lambda df: df.groupby('a').sum()").withIndexes())
+```
+
+**Using the PythonExternalTransform class**
+
+When an SDK-specific wrapper is not available, you cound access the Python cross-language transform through the `PythonExternalTransform` class by specifying the fully qualified name and the constructor arguments of the target Python transform.
```
Comment (pcoet) [inline website/www/site/content/en/documentation/programming-guide.md:7419]: "cound" -> "can"

**Review by pcoet — APPROVED** [review]
LGTM. Suggested a lightly edited alternative to one paragraph.

File: website/www/site/content/en/documentation/programming-guide.md (hunk 1 above)
Comment (ihji) [reply website/www/site/content/en/documentation/programming-guide.md]: Done.

File: website/www/site/content/en/documentation/programming-guide.md (hunk 2 above)
Comment (ihji) [reply website/www/site/content/en/documentation/programming-guide.md]: Done.

**Comment by ihji** [comment]
@pcoet Thanks!

File: website/www/site/content/en/documentation/programming-guide.md [hunk 3]
```diff
@@ -7282,7 +7282,16 @@ To create an SDK wrapper for use in a Python pipeline, do the following:
 
 #### 13.1.2. Creating cross-language Python transforms
 
-To make your Python transform usable with different SDK languages, you must create a Python module that registers an existing Python transform as a cross-language transform for use with the Python expansion service and calls into that existing transform to perform its intended operation.
+Any Python transforms defined in the scope of the expansion service should be accessible by specifying their fully qualified names. For example, you could use Python's `ReadFromText` transform in a Java pipeline with its fully qualified name `apache_beam.io.ReadFromText`:
+
+```java
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/programming-guide.md:7287]: Can we just add Go snippets to also cover this part for Go SDK or will it need other documentation changes ?

cc: @riteshghorse

File: website/www/site/content/en/documentation/programming-guide.md [hunk 4]
```diff
@@ -7282,7 +7282,16 @@ To create an SDK wrapper for use in a Python pipeline, do the following:
...
+p.apply("Read",
+    PythonExternalTransform.<PBegin, PCollection<String>>from("apache_beam.io.ReadFromText")
+    .withKwarg("file_pattern", options.getInputFile())
+    .withKwarg("validate", false))
+```
+
+Alternatively, you may want to create a Python module that registers an existing Python transform as a cross-language transform for use with the Python expansion service and calls into that existing transform to perform its intended operation. A registered URN can be used later in an expansion request for indicating an expansion target.
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/programming-guide.md:7294]: Should we provide an example snippet for this as well ?

File: website/www/site/content/en/documentation/programming-guide.md [hunk 5]
```diff
@@ -7393,7 +7402,29 @@ Depending on the SDK language of the pipeline, you can use a high-level SDK-wrap
...
+Users have three options to use cross-language transforms in a Java pipeline. At the highest level of abstraction, some popular Python transforms are accessible through dedicated Java wrapper transforms. For example, the Java SDK has the `DataframeTransform` class, which uses the Python SDK's `DataframeTransform`, and it has the `RunInference` class, which uses the Python SDK's `RunInference`, and so on. When an SDK-specific wrapper transform is not available for a target Python transform, you can use the lower-level [PythonExternalTransform](https://github.com/apache/beam/blob/master/sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/PythonExternalTransform.java) class instead by specifying the fully qualified name of the Python transform. If you want to try external transforms from SDKs other than Python, you can also use the lowest-level [External](https://github.com/apache/beam/blob/master/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java) class.
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/programming-guide.md:7405]: We should try to make this less language combination specific (Python from Java) if possible. Probably structure it as:

\#\#\#\#\# Simplified APIs for using Python transforms from Java
\#\#\#\#\# Generic API for using arbitrary transforms from Java 

Also, please mention that latter applies for Java-on-Java as well.

File: website/www/site/content/en/documentation/programming-guide.md [hunk 6]
```diff
@@ -7282,7 +7282,16 @@ To create an SDK wrapper for use in a Python pipeline, do the following:
 
 #### 13.1.2. Creating cross-language Python transforms
 
-To make your Python transform usable with different SDK languages, you must create a Python module that registers an existing Python transform as a cross-language transform for use with the Python expansion service and calls into that existing transform to perform its intended operation.
+Any Python transforms defined in the scope of the expansion service should be accessible by specifying their fully qualified names. For example, you could use Python's `ReadFromText` transform in a Java pipeline with its fully qualified name `apache_beam.io.ReadFromText`:
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/programming-guide.md:7285]: This still requires "from_runner_api_proto" " "to_runner_api_proto" methods to be available in the Python transform, correct ?

If so we should note that here.

**Review by chamikaramj — COMMENTED** [review]
Thanks.

File: website/www/site/content/en/documentation/programming-guide.md (hunk 3 above)
Comment (ihji) [reply website/www/site/content/en/documentation/programming-guide.md:7287]: I think it deserves to be a separate PR.

File: website/www/site/content/en/documentation/programming-guide.md [hunk 7]
```diff
@@ -7282,7 +7282,16 @@ To create an SDK wrapper for use in a Python pipeline, do the following:
 
 #### 13.1.2. Creating cross-language Python transforms
```
Comment (chamikaramj) [inline website/www/site/content/en/documentation/programming-guide.md:7284]: Can you also update:

(1) "Including dependencies" section, to include information regarding "PythonExternalTransform.withExtraPackages" option.

(2) "Starting the expansion service" section to include "--pickle_library=cloudpickle" option. 

This should include

**Review by chamikaramj — COMMENTED** [review]
Thanks. LGTM.

File: website/www/site/content/en/documentation/programming-guide.md (hunk 3 above)
Comment (riteshghorse) [reply website/www/site/content/en/documentation/programming-guide.md:7287]: Ok, I'll track it in a separate issue. Thanks!

File: website/www/site/content/en/documentation/programming-guide.md (hunk 7 above)
Comment (ihji) [reply website/www/site/content/en/documentation/programming-guide.md:7284]: Done.

**Pull request merged**